sdk/aqueduct/utils/naming.py

```diff
@@ -0,0 +1,101 @@
+from typing import List, Optional, Tuple, Union
+
+from aqueduct.models.dag import DAG
+from aqueduct.utils.dag_deltas import RemoveOperatorDelta, apply_deltas_to_dag
+
+
+def _artifact_name_from_op_name(op_name: str) -> str:
+    return op_name + " artifact"
+
+
+def _construct_default_output_artifact_names_from_op(op_name: str, num_outputs: int) -> List[str]:
+    """The default artifact naming policy is "<op_name> artifact (<optional counter>)".
+
+    In the multi-output case, we deduplicate each, starting the counter at 1.
+    """
+    artifact_names = [_artifact_name_from_op_name(op_name)]
+    if num_outputs == 1:
+        return artifact_names
+
+    artifact_names += [
+        _artifact_name_from_op_name(op_name) + " (%d)" % i for i in range(1, num_outputs)
+    ]
+    return artifact_names
+
+
+def resolve_op_and_artifact_names(
+    dag: DAG,
+    candidate_op_name: str,
+    overwrite_existing_op_name: bool,
+    candidate_artifact_names: Optional[Union[str, List[str]]] = None,
+    num_outputs: int = 1,
+    only_resolve_op_name: bool = False,
+) -> Tuple[str, List[str]]:
+    """Enforces our unique naming policy on operators and also their output artifacts.
+
+    Operator collisions are resolved. Artifact collisions are not tolerated.
+
+    Args:
+        dag:
+            The current dag.
+        candidate_op_name:
+            The operator name that we want to check and potentially deduplicate.
+        overwrite_existing_op_name:
+            If set, any operator collisions will be ignored. Otherwise, we will bump
+            the operator name until an available one is found.
+        candidate_artifact_names:
+            The custom output artifact names the caller wants to assign to the operator
+            outputs. If not set, we will use our default artifact naming scheme. Regardless,
+            if there already exist artifact's with the same name, we will error.
+        num_outputs:
+            Must be consistent with `candidate_artifact_names`, if set. The number of
+            output artifacts to expected.
+        only_resolve_op_name:
+            If set, skip the artifact naming validation step. An empty list will be
+            returned for the artifact names in that case.
+    Returns:
+        A tuple of the operator name and artifact name(s), which the caller can use
+        to safely attach elements to the DAG.
+    """
+    # First, resolve the candidate op name. If we can overwrite an existing op name, then
+    # don't even bother finding an unallocated op name.
+    if not overwrite_existing_op_name:
+        prefix = candidate_op_name
+        curr_suffix = 1
+        while dag.get_operator(with_name=candidate_op_name) is not None:
+            candidate_op_name = prefix + " (%d)" % curr_suffix
+            curr_suffix += 1
+
+    # If we're overwriting an existing operator, temporarily delete the operator from the dag
```

Comment:
Not sure I totally understand this part. Are we expecting the caller to add the operator with the same name to the DAG later on? Also just to confirm, when we delete this operator, all of its downstream ops are also deleted right?

Reply:
I think my understanding is correct after reading through the whole PR. Maybe worth adding a comment clarifying the contract here?

Reply:
No, this is so that the artifact naming collision detection is accurate. If I don't delete the operator I'm colliding with, then it's artifact may collide with the candidate artifact in the second half of this method. Extending this logic, any artifact downstream of this existing operator needs to be removed before we can compute artifact name uniqueness.

Reply:
I think what you're saying make sense, but still within this function, we aren't really "temporarily" deleting the operator because we're not adding it back later within this function right? So it's still up to the caller to add the candidate operator to this DAG?